falcon/asgi/ws.py

```diff
@@ -0,0 +1,671 @@
+import asyncio
+import collections
+from enum import Enum
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    Deque,
+    Dict,
+    Iterable,
+    Mapping,
+    Optional,
+    Union,
+)
+
+import falcon
+from falcon import errors, media
+from falcon.asgi_spec import EventType, WSCloseCode
+from falcon.constants import WebSocketPayloadType
+
+
+_WebSocketState = Enum('_WebSocketState', 'HANDSHAKE ACCEPTED CLOSED')
+
+
+__all__ = ['WebSocket']
+
+
+class WebSocket:
+    """Represents a single WebSocket connection with a client.
+
+    Attributes:
+        ready (bool): ``True`` if the WebSocket connection has been
+            accepted and the client is still connected, ``False`` otherwise.
+        unaccepted (bool)): ``True`` if the WebSocket connection has not yet
+            been accepted, ``False`` otherwise.
+        closed (bool): ``True`` if the WebSocket connection has been closed
+            by the server or the client has disconnected.
+        subprotocols (tuple[str]): The list of subprotocol strings advertised
+            by the client, or an empty tuple if no subprotocols were
+            specified.
+        supports_accept_headers (bool): ``True`` if the ASGI server hosting
+            the app supports sending headers when accepting the WebSocket
+            connection, ``False`` otherwise.
+
+    """
+
+    __slots__ = (
+        '_asgi_receive',
+        '_asgi_send',
+        '_buffered_receiver',
+        '_close_code',
+        '_supports_accept_headers',
+        '_mh_bin_deserialize',
+        '_mh_bin_serialize',
+        '_mh_text_deserialize',
+        '_mh_text_serialize',
+        '_state',
+        'subprotocols',
+    )
+
+    def __init__(
+        self,
+        ver: str,
+        scope: dict,
+        receive: Callable[[], Awaitable[dict]],
+        send: Callable[[dict], Awaitable],
+        media_handlers: Mapping[
+            WebSocketPayloadType,
+            Union[media.BinaryBaseHandlerWS, media.TextBaseHandlerWS]
+        ],
+        max_receive_queue: int,
+    ):
+        self._supports_accept_headers = (ver != '2.0')
+
+        # NOTE(kgriffs): Normalize the iterable to a stable tuple; note that
+        #   ordering is significant, and so we preserve it here.
+        self.subprotocols = tuple(scope.get('subprotocols', []))
+
+        # TODO(kgriffs): Should we make the use of _BufferedReceiver
+        #   configurable? For example, if the developer knows that
+        #   they will be interleaving receives and sends, then they
+        #   will be able to find out about the 'websocket.disconnect'
+        #   event via one of their receive() calls, and there is no
+        #   need for the added overhead.
+        #
+        #   On the other hand, using _BufferedReceiver may still be
```

Comment:
``_BufferedReceiver`` probably can't assist with back pressure after all, since Falcon is just acting in the stead of the server app, and when Falcon stops receiving messages due to the queue being full, the WebSocket server will just keep receiving from the client in a separate async task.